service/accessor.go

```diff
@@ -96,3 +100,237 @@ func (ua *UpdatableAccessor) Update(a Accessor, err error) {
 	ua.current = a
 	ua.lock.Unlock()
 }
+
+type RouteTraffic interface {
+	Route(instance string) error
+}
+
+type emptyRouter struct {
+}
+
+func (r emptyRouter) Route(instance string) error {
+	return nil
+}
+
+func DefaultTrafficRouter() RouteTraffic {
+	return emptyRouter{}
+}
+
+type AccessorQueue interface {
+	Order([]string) []string
+}
+
+type emptyQueue struct {
+	r *rand.Rand
+}
+
+func DefaultOrder() AccessorQueue {
+	return emptyQueue{}
+}
+
+func (r emptyQueue) Order(keys []string) []string {
+	return keys
+}
+
+type AccessorValue struct {
+	Accessor Accessor
+	Err      error
+}
+
+type LayeredAccessor struct {
```

Comment:
This should really be an interface, with an internal implementation such as `layeredAccessor`. 

Currently, this type is mutable, which means you have to have a mutex for things like the router.  By using an interface returned by a `NewLayeredAccessor` function, the returned `LayeredAccessor` can be immutable.  This, in turn, removes any chance for a deadlock now or in the future.